### PR TITLE
:warning: split the webhook builder out as a separate builder.

### PR DESCRIPTION
**Pull request by mengqiy**
`build.go`=> `controller.go` + `webhook.go`
`build_test.go`=> `controller_test.go` + `webhook_test.go`

TODO: figure out if `For` method in webhook builder should be variadic?

### CODE REVIEW

File: pkg/builder/webhook.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/conversion"
+)
+
+// WebhookBuilder builds a Controller.
```
Comment (droot) [inline pkg/builder/webhook.go]: s/Controller/Webhook

File: pkg/builder/webhook.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+// If the passed in object has implemented the admission.Defaulter interface, a MutatingWebhook will be wired for this type.
+// If the passed in object has implemented the admission.Validator interface, a ValidatingWebhook will be wired for this type.
+// TODO: figure out if we support only one object or multiple objects
+func (blder *WebhookBuilder) For(apiType runtime.Object) *WebhookBuilder {
+	blder.apiType = apiType
+	return blder
+}
+
+// Complete builds the Application ControllerManagedBy.
+func (blder *WebhookBuilder) Complete() error {
```
Comment (DirectXMan12) [inline pkg/builder/webhook.go]: do we really need this?  Can't we just do the equivalent in `For`?

File: pkg/builder/webhook.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+	// Set the Webook if needed
+	if err := blder.doWebhook(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (blder *WebhookBuilder) doConfig() error {
```
Comment (DirectXMan12) [inline pkg/builder/webhook.go]: this should have a better name than `doConfig` -- perhaps `loadRestConfig`?

File: pkg/builder/webhook.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	return &WebhookBuilder{mgr: m}
+}
+
+// For defines the type of Object being *reconciled*, and configures the ControllerManagedBy to respond to create / delete /
+// update events by *reconciling the object*.
+// This is the equivalent of calling
+// Watches(&source.Kind{Type: apiType}, &handler.EnqueueRequestForObject{})
+// If the passed in object has implemented the admission.Defaulter interface, a MutatingWebhook will be wired for this type.
+// If the passed in object has implemented the admission.Validator interface, a ValidatingWebhook will be wired for this type.
+// TODO: figure out if we support only one object or multiple objects
```
Comment (droot) [inline pkg/builder/webhook.go]: I think we can simplify this comment now.

File: pkg/builder/webhook.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	if blder.mgr != nil {
+		blder.config = blder.mgr.GetConfig()
+		return nil
+	}
+	var err error
+	blder.config, err = getConfig()
+	return err
+}
+
+func (blder *WebhookBuilder) doManager() error {
```
Comment (DirectXMan12) [inline pkg/builder/webhook.go]: not needed -- was only in the old builder for legacy reasons (before we switched to `ManagedBy`)

File: pkg/builder/webhook.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	if blder.mgr != nil {
+		return nil
+	}
+	var err error
+	blder.mgr, err = newManager(blder.config, manager.Options{})
+	return err
+}
+
+func (blder *WebhookBuilder) doWebhook() error {
```
Comment (DirectXMan12) [inline pkg/builder/webhook.go]: ditto here -- maybe `registerRelatedWebhooks` or something?

File: pkg/builder/webhook.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	// 2) convert it to the hub version
+	// 3) do defaulting
+	// 4) convert it back to the same bespoke version
+	// 5) calculate the JSON patch
+	//
+	// A potential workflow for validating webhook
+	// 1) a bespoke (non-hub) version comes in
+	// 2) convert it to the hub version
+	// 3) do validation
+	if defaulter, isDefaulter := blder.apiType.(admission.Defaulter); isDefaulter {
```
Comment (DirectXMan12) [inline pkg/builder/webhook.go]: split these into helpers

**Comment by DirectXMan12** [comment]
Need to split out conversion registration as well (in a separate PR) @droot

File: pkg/builder/webhook.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+			if !blder.isAlreadyHandled(path) {
+				log.Info("Registering a validating webhook",
+					"GVK", gvk,
+					"path", path)
+				blder.mgr.GetWebhookServer().Register(path, vwh)
+			}
+		}
+	}
+
+	err = conversion.CheckConvertibility(blder.mgr.GetScheme(), blder.apiType)
```
Comment (droot) [inline pkg/builder/webhook.go]: I will make a separate PR on top of this to enable conversion only when type implements convertible.

File: pkg/builder/webhook.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+	// 1) a bespoke (non-hub) version comes in
+	// 2) convert it to the hub version
+	// 3) do defaulting
+	// 4) convert it back to the same bespoke version
+	// 5) calculate the JSON patch
+	//
+	// A potential workflow for validating webhook
+	// 1) a bespoke (non-hub) version comes in
+	// 2) convert it to the hub version
+	// 3) do validation
```
Comment (droot) [inline pkg/builder/webhook.go]: Do we need the above comments ?

**Comment by mengqiy** [comment]
PTAL

File: pkg/builder/webhook.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	config  *rest.Config
+}
+
+func WebhookManagedBy(m manager.Manager) *WebhookBuilder {
+	return &WebhookBuilder{mgr: m}
+}
+
+// TODO(droot): update the GoDoc for conversion.
+
+// For takes a runtime.Object which should be a CR.
```
Comment (droot) [inline pkg/builder/webhook.go]: Not necessarily this PR, but we should consider adding an example for this that shows in godoc. They are super useful.

Comment (mengqiy) [reply pkg/builder/webhook.go]: Done!

Comment (mengqiy) [reply pkg/builder/webhook.go]: https://github.com/kubernetes-sigs/controller-runtime/pull/497/files#diff-8b3e76472aacd449b1d1dda72eefbc32

**Comment by mengqiy** [comment]
PTAL

File: pkg/builder/webhook.go (hunk 2 above)
Comment (droot) [reply pkg/builder/webhook.go]: I realized without having `Complete()` method, the build chain looks incomplete and its not clear whether we are done with the builder or not. So I would suggest not dropping this method. It also keeps it consistent with the controller builder.

Comment (mengqiy) [reply pkg/builder/webhook.go]: We decided to bring `Complete` back.

**Pull request merged**